File: day03/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

var partOneRegex = regexp.MustCompile(`mul\((\d{1,3}),(\d{1,3})\)`)
var partTwoRegex = regexp.MustCompile(`do\(\)|don't\(\)|mul\((\d{1,3}),(\d{1,3})\)`)

func readLines(filename string) []string {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		panic(err)
	}
	return lines
}

func part1(lines []string) {
	// Compile the regex once

	sum := 0

	for _, line := range lines {
		// Find all matches in the line
		matches := partOneRegex.FindAllStringSubmatch(line, -1)
		for _, match := range matches {
			// Convert the matched numbers to integers
			first, _ := strconv.Atoi(match[1])
			second, _ := strconv.Atoi(match[2])
			sum += first * second
		}
	}

	fmt.Println("Part 1:", sum)
}

func part2(lines []string) {
	// Compile the regex once
	mulEnabled := true
	sum := 0

	for _, line := range lines {
		// Find all matches in the line
		matches := partTwoRegex.FindAllStringSubmatch(line, -1)
		for _, match := range matches {
			switch match[0] {
			case "do()":
				mulEnabled = true
			case "don't()":
				mulEnabled = false
			default: // Handle "mul(x, y)" cases
				if mulEnabled {
					first, _ := strconv.Atoi(match[1])
					second, _ := strconv.Atoi(match[2])
					sum += first * second
				}
			}
		}
	}

	fmt.Println("Part 2:", sum)
}

func main() {
	lines := readLines("input.txt")
	part1(lines)
	part2(lines)
}

File: day03/main.py
import re


def get_input(test: str = None):
    with open('./input.txt') as f:
        base = test or f.read()
        l = base.splitlines()

    return l


def part_1(l: list[str]):
    result = 0

    for line in l:
        valid = re.findall(r'mul\((\d+),(\d+)\)', line)

        for x, y in valid:
            result += int(x) * int(y)

    print(result)


def part_2(l: list[str]):
    result = 0
    text = ''.join(l)

    while True:
        i = text.find("don't()")
        valid = re.findall(r'mul\((\d+),(\d+)\)', text[:i])

        for x, y in valid:
            result += int(x) * int(y)

        text = text[i:]
        j = text.find("do()")
        text = text[j + 4 :]

        if j == -1:
            break

    print(result)


part_1(get_input())
part_2(get_input())